src/complete.c

```diff
@@ -169,12 +170,31 @@ void list_detect_complete(SEXP x, R_len_t size, int* p_out) {
 
 // -----------------------------------------------------------------------------
 
+static inline void col_detect_complete_switch(SEXP x, R_len_t size, int* p_out);
+
 static inline
 void df_detect_complete(SEXP x, R_len_t size, int* p_out) {
   r_ssize n_cols = r_length(x);
   const SEXP* p_x = VECTOR_PTR_RO(x);
 
   for (r_ssize i = 0; i < n_cols; ++i) {
-    vec_detect_complete_switch(p_x[i], size, p_out);
+    col_detect_complete_switch(p_x[i], size, p_out);
+  }
+}
+
+static inline
+void col_detect_complete_switch(SEXP x, R_len_t size, int* p_out) {
```

Comment:
It's a copy of `vec_detect_complete_switch()` with different branches for data frames and `NULL`

It mirrors how we do it in `vec_detect_missing()`